provider/cloud/baremetal/baremetal.go

```
@@ -0,0 +1,204 @@
+package baremetal
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"path"
+
+	"github.com/ghodss/yaml"
+	"github.com/golang/glog"
+	"github.com/kubermatic/api"
+	"golang.org/x/net/context"
+)
+
+const serviceURL = "baremetal-provider.api.svc.cluster.local"
+const appJSON = "application/json"
+
+type baremetal struct {
+}
+
+func (*baremetal) InitializeCloudSpec(cl *api.Cluster) error {
+	cfg := cl.GetKubeconfig()
+
+	jcfg, err := json.Marshal(cfg)
+	if err != nil {
+		return err
+	}
+
+	ycfg, err := yaml.JSONToYAML(jcfg)
+	if err != nil {
+		return err
+	}
+	cl.Spec.Cloud.BareMetal = &api.BareMetalCloudSpec{
+		Name:         cl.Metadata.Name,
+		APIServerURL: cl.Address.URL,
+		KubeConfig:   string(ycfg),
+	}
+
+	url := path.Join(serviceURL, "/clusters")
+	data, err := json.Marshal(cl.Spec.Cloud.BareMetal)
+	if err != nil {
+		return err
+	}
+
+	_, err = http.Post(url, appJSON, bytes.NewReader(data))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (*baremetal) MarshalCloudSpec(cls *api.CloudSpec) (annotations map[string]string, err error) {
+	annotations = map[string]string{
+		"apiserver_url": cls.BareMetal.APIServerURL,
+		"name":          cls.BareMetal.Name,
+		"kubeconfig":    cls.BareMetal.KubeConfig,
+	}
+	return annotations, nil
+}
+
+func (*baremetal) UnmarshalCloudSpec(annotations map[string]string) (*api.CloudSpec, error) {
+	var cl *api.CloudSpec
+
+	url, ok := annotations["apiserver_url"]
+	if !ok {
+		return nil, errors.New("couldn't find key")
+	}
+	cl.BareMetal.APIServerURL = url
+
+	name, ok := annotations["name"]
+	if !ok {
+		return nil, errors.New("couldn't find key")
```

Please return a meaningful error message which also includes the missing key